client/src/includes/breadcrumbs.test.js

```diff
@@ -29,7 +29,7 @@ describe('initCollapsibleBreadcrumbs', () => {
       <span id="username" data-breadcrumb-item />
     </nav>
   </div>
-</div>`;
+</header>`;
```

Comment:
Can we please add a test in this file for the data attribute being supplied also.

Reply:
A test for checking if `dataset-selected-header=true` is present? Or do you mean to write a test expecting nothing when `dataset-selected-header=true` is not present?

Reply:
That's not how it's meant to work. 

Basically we can pass in any selector to the data attribute and it should find the closest parent of that. 

Do if we have data-header-selector="#some-id'

It should find the nearest parent with the id 'some-id' and when that gets the hover / mouse leave it should driver the behaviour. As in the event listener should be attached to that element

Reply:
I searched everywhere but couldn't find any way to trigger `mouseenter` and `mouseleave` events using javascript. Found one approach using jquery i.e. using `$('#id').mouseenter()` but it has been deprecated since Jquery 3.3.

Reply:
There are two approaches here, one is to use the existing test's spy and write some more specific checks, the other is  to fake the call with `document.getElementById('some-id).dispatch(new Event('mouseenter', {});`.

## Option one  - more specific checks against the spy.

```
  const spy = jest.spyOn(document, 'addEventListener');
```

This `spy` (maybe we should rename it to `addEventListenerSpy`) will spy on all calls to addEventListener.

Our goal here is to check whether or not the right element gets the event listeners added, so in the other test you could write something that checks _what_ element `addEventListenerSpy` was listened to via `expect(addEventListenerSpy).toHaveBeenCalledWith(something)`, the something being the expected element with the id.

You can also get the mock (aka spy) function calls  via [`addEventListenerSpy.mock.calls`](https://jestjs.io/docs/mock-function-api#mockfnmockcalls).

## Option two - dispatch events

In JSDOM (which is the fake browser environment that is set up in your Jest tests) you can access most browser APIs, one of them is `dispatchEvent` https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent

This lets you dispatch an event with any name, including browser built in ones (as the name is just a string). There are some small nuances here but for the sake of a test they do not matter.

## Finally, how to approach this all.

Have a think about what we are trying to do;

* We have a data attribute that allows the code to work differently
* That data attribute `data-header-selector` is a query selector and gets passed into [element.closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)
* So start with the HTML in your new test, write some HTML first